packages/iov-bns/src/stream.ts

```diff
@@ -0,0 +1,80 @@
+import { InternalListener, InternalProducer, Stream } from "xstream";
+
+// streamPromise takes a Promsie that returns an array
+// and emits one event for each element of the array
+export function streamPromise<T>(promise: Promise<ReadonlyArray<T>>): Stream<T> {
+  return new Stream<T>(new StreamPromise<T>(promise));
+}
+
+export async function readIntoArray<T>(stream: Stream<T>): Promise<ReadonlyArray<T>> {
+  const concat = (acc: ReadonlyArray<T>, val: T) => acc.concat(val);
+  const result = stream.fold(concat, []).last();
+  return new Promise<ReadonlyArray<T>>((resolve, reject) => {
+    result.addListener({
+      next: (val: ReadonlyArray<T>) => resolve(val),
+      error: (err: any) => reject(err),
+    });
+  });
+}
+
+export async function countStream<T>(stream: Stream<T>): Promise<number> {
+  // tslint:disable-next-line:no-let
+  let count = 0;
+  return new Promise<number>((resolve, reject) => {
+    stream.addListener({
+      next: () => {
+        count++;
+      },
+      error: (err: any) => reject(err),
+      complete: () => resolve(count),
```

Comment:
feels like this can only count streams (in the future) of length 0 or 1?  i noticed in the test that counts events indeed one is historical... so an easy way to test the hypothesis is to make 2 events in the future.  if that is already tested and i missed it please let me know otherwise i think this is another case where it looks like it eats all future events after the first to me.

Reply:
A stream will call next() N times.
It will end with one call to either error or complete.

If there are 5 events, they will call next 5 times, then complete upon closing the stream (why i have an explicit disconnect at the end of the watch tests).

Test case here: https://github.com/iov-one/iov-core/blob/bns-subscriptions/packages/iov-bns/src/stream.spec.ts#L23-L29

Reply:
yeah i think you are right in this explanation you gave now that I reread the code @ethanfrey  however the problem is that the countStream test still seems to return `2` when i expect `3` in the other pr based on this one with a new event added. so until that is explained we can block this to figure out where the mistake is.